source/common/io/io_uring_worker.h

```diff
@@ -0,0 +1,509 @@
+                      write_buf_.length(), fd_);
+            submitWriteRequest();
+          }
+          return;
+        }
+      }
+
+      WriteParam param{result};
+      io_uring_handler_.onWrite(param);
+      return;
+    }
+
+    write_buf_.drain(result);
+    ENVOY_LOG(trace, "drain write buf, drain size = {}, fd = {}", result, fd_);
+    if (write_buf_.length() > 0) {
+      ENVOY_LOG(trace, "continue write buf since still have data, size = {}, fd = {}",
+                write_buf_.length(), fd_);
+      submitWriteRequest();
+      return;
+    }
+
+    if (is_closing_ && read_req_ == nullptr && cancel_req_ == nullptr) {
+      ENVOY_LOG(trace, "write is done, try to close the socket, fd = {}", fd_);
+      ASSERT(close_req_ == nullptr);
+      close_req_ = parent_.submitCloseRequest(*this);
+      return;
+    }
+
+    if (is_full_ && write_buf_.length() == 0) {
+      is_full_ = false;
+      WriteParam param{result};
+      io_uring_handler_.onWrite(param);
+    }
+  }
+
+  uint64_t write(Buffer::Instance& buffer) override {
+    ENVOY_LOG(trace, "write, size = {}, fd = {}, slices = {}", buffer.length(), fd_,
+              buffer.getRawSlices().size());
+    uint64_t buffer_limit = 1024 * 1024;
+    uint64_t writable_size = buffer_limit - write_buf_.length();
+    if (writable_size <= 0) {
+      is_full_ = true;
+      return 0;
+    }
+    auto length_to_write = std::min(buffer.length(), writable_size);
+    write_buf_.move(buffer, length_to_write);
+
+    if (write_req_ != nullptr) {
+      ENVOY_LOG(trace, "write already submited, fd = {}", fd_);
+      return length_to_write;
+    }
+
+    ENVOY_LOG(trace, "write buf, size = {}, slices = {}", write_buf_.length(),
+              write_buf_.getRawSlices().size());
+    submitWriteRequest();
+    return length_to_write;
+  }
+
+  uint64_t writev(const Buffer::RawSlice* slices, uint64_t num_slice) override {
+    ENVOY_LOG(trace, "writev, num_slices = {}, fd = {}", num_slice, fd_);
+
+    uint64_t data_to_write = 0;
+    for (uint64_t i = 0; i < num_slice; i++) {
+      write_buf_.add(slices[i].mem_, slices[i].len_);
+      data_to_write += slices[i].len_;
+    }
+
+    if (write_req_ != nullptr) {
+      ENVOY_LOG(trace, "write already submited, fd = {}", fd_);
+      return data_to_write;
+    }
+
+    submitWriteRequest();
+    return data_to_write;
+  }
+
+protected:
+  void submitReadRequest() {
+    ENVOY_LOG(trace, "submit read request, fd = {}", fd_);
+    ASSERT(read_req_ == nullptr);
+    ASSERT(iouring_read_buf_ == nullptr);
+    iouring_read_buf_ = std::make_unique<uint8_t[]>(read_buffer_size_);
+    iov_->iov_base = iouring_read_buf_.get();
+    iov_->iov_len = read_buffer_size_;
+    read_req_ = parent_.submitReadRequest(*this, iov_);
+  }
+
+  void submitWriteRequest() {
+    ASSERT(write_req_ == nullptr);
+    ASSERT(iovecs_ == nullptr);
+
+    Buffer::RawSliceVector slices = write_buf_.getRawSlices();
+    iovecs_ = new struct iovec[slices.size()];
```

Comment:
What's the reason we need to use new directly here?

Reply:
emm... no, I should be able to use unique ptr here.